### PR TITLE
CC-KBC | SGX Occlum Attester

### PR DESCRIPTION
**Pull request by Xynnn007**
> **warning** CI machine is not prepared, related scripts and yamls will not work

### CODE REVIEW

**Comment by Xynnn007** [comment]
@qzheng527 @mythi  Hi, please take a look if interested and check whether I'm doing something stupid. We can support cc-kbc & kbs after v0.5.0 in enclave-cc.

**Comment by jepio** [comment]
In `cc_kbc.yml` you will want to enable the tdx feature together with cc_kbc, yes?

**Comment by Xynnn007** [comment]
> In `cc_kbc.yml` you will want to enable the tdx feature together with cc_kbc, yes?

Yes. What I want is to bring `tdx` and other attesters out of `cc-kbc` as new features to let user decide which attester to use (Still needs review, which is discussed in https://github.com/confidential-containers/guest-components/issues/212) The last line is only a lint test, so I enabled all the attester features (currently tdx and occlum in sgx)

Thanks for informing, I found something wrong in the app Cargo.toml. Let me fix this

File: .github/workflows/occlum_sgx.yml [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    paths:
+      - 'src/kbc_modules/cc_kbc/attesster/sgx_occlum'
+      - '.github/workflows/occlum_sgx.yml'
+
+jobs:
+  build-and-run-occlum:
+    runs-on: self-hosted
+    container:
+      image: occlum/occlum:latest-ubuntu20.04
+      options: --device /dev/sgx/enclave --device /dev/sgx/provision
```
Comment (mythi) [inline .github/workflows/occlum_sgx.yml:21]: does the runner use some old driver or can you go with:
```suggestion
      options: --device /dev/sgx_enclave --device /dev/sgx_provision
```

File: ci/occlum.sh [hunk 2]
```diff
@@ -0,0 +1,10 @@
+rm -rf occlum_instance && mkdir occlum_instance && cd occlum_instance
+
+occlum init && rm -rf image
+
+# Set the kernel space heap size
+sed -i 's/\"kernel_space_heap_size\": \"32MB\"/\"kernel_space_heap_size\": \"100MB\"/g' Occlum.json;
```
Comment (mythi) [inline ci/occlum.sh:6]: FWIW, I've found `jq` filters easier to manage.

File: .github/workflows/occlum_sgx.yml [hunk 3]
```diff
@@ -0,0 +1,40 @@
+name: CC kbc build CI
+on:
+  push:
+    paths:
+      - 'src/kbc_modules/cc_kbc/attesster/sgx_occlum'
```
Comment (mythi) [inline .github/workflows/occlum_sgx.yml:5]:
```suggestion
      - 'src/kbc_modules/cc_kbc/attester/sgx_occlum'
```

File: .github/workflows/occlum_sgx.yml [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+      - '.github/workflows/occlum_sgx.yml'
+  pull_request:
+    paths:
+      - 'src/kbc_modules/cc_kbc/attesster/sgx_occlum'
```
Comment (mythi) [inline .github/workflows/occlum_sgx.yml:9]:
```suggestion
      - 'src/kbc_modules/cc_kbc/attester/sgx_occlum'
```

File: .github/workflows/occlum_sgx.yml [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+      - '.github/workflows/occlum_sgx.yml'
+  create:
+    paths:
+      - 'src/kbc_modules/cc_kbc/attesster/sgx_occlum'
```
Comment (mythi) [inline .github/workflows/occlum_sgx.yml:13]:
```suggestion
      - 'src/kbc_modules/cc_kbc/attester/sgx_occlum'
```

**Review by mythi — COMMENTED** [review]
just some small nits.

does this mean we will move away from RATS-TLS dependency in enclave-cc/agent-enclave?

File: .github/workflows/occlum_sgx.yml (hunk 1 above)
Comment (Xynnn007) [reply .github/workflows/occlum_sgx.yml:21]: I found this in occlum's document https://github.com/occlum/occlum#how-to-use, step 3. I manually create the soft link `/dev/sgx/enclave` from `/dev/sgx_enclave`. I'm not sure whether it is for occlum? cc @qzheng527

File: ci/occlum.sh (hunk 2 above)
Comment (Xynnn007) [reply ci/occlum.sh:6]: Great suggestion, let me fix this.

**Comment by Xynnn007** [comment]
@mythi Thanks a lot for your great suggestions!
> does this mean we will move away from RATS-TLS dependency in enclave-cc/agent-enclave?
- If we use eaa-kbc & verdictd, we still need RATS-TLS dep in enclave-cc.
- If we use cc-kbc & [cc-kbs](https://github.com/confidential-containers/kbs), we do not need RATS-TLS in enclave-cc.
I think we can make them into different features and by default (?) cc-kbc suites.

cc @haosanzi 

Also, sgx runners for the repo will be ready soon (shared on with enclave-cc). Thanks for help from @fitzthum

File: .github/workflows/occlum_sgx.yml (hunk 1 above)
Comment (mythi) [reply .github/workflows/occlum_sgx.yml:21]: Occlum uses SGX SDK underneath and the latest SDK versions support the standard in-tree kernel provided device nodes. The symlinks are not needed and to stay future proof, `/dev/sgx_enclave` path is preferred.

**Review by jialez0 — APPROVED** [review]

**Pull request merged**